### PR TITLE
Scalar Initial Condition

### PR DESCRIPTION
**Pull request by timfelle**
- Added support for initial conditions in the scalar field
- Updated documentation with changes in the case file
- Default behavior is retained as uniform field with value 0

### CODE REVIEW

**Review by njansson — COMMENTED** [review]
Very nice!
How about updating the RBC example and/or the reframe test case to use scalar ic instead of setting the initial scalar values via `fluid_user_ic`?

**Comment by timfelle** [comment]
Sure I can take a look, didn't think any of the examples used it since it was not implemented.

**Comment by njansson** [comment]
> Sure I can take a look, didn't think any of the examples used it since it was not implemented.

Great thanks!

File: reframe/src/rayleigh.f90 [hunk 1]
```diff
@@ -50,27 +50,18 @@ subroutine scalar_bc(s, x, y, z, nx, ny, nz, ix, iy, iz, ie, t, tstep)
...
+    type(field_t), intent(inout) :: s
     type(json_file), intent(inout) :: params
-    type(field_t), pointer :: s
     integer :: i, e, k, j
     real(kind=rp) :: rand, z
-    s => neko_field_registry%get_field('s')
 
-    call rzero(u%x,u%dof%size())
-    call rzero(v%x,v%dof%size())
-    call rzero(w%x,w%dof%size())
```
Comment (njansson) [inline reframe/src/rayleigh.f90:72]: We need to retain this (I think), thus having both a `fluid_user_ic` and a `scalar_user_ic`

Comment (timfelle) [reply reframe/src/rayleigh.f90:72]: I changed the case file to just use a uniform initial condition for the fluid, so it should be good.

File: src/scalar/scalar_ic.f90 [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+       call device_memcpy(s%x, s%x_d, s%dof%size(), HOST_TO_DEVICE)
+    end if
+
+    ! Ensure continuity across elements for initial conditions
+    call gs%op(s%x, s%dof%size(), GS_OP_ADD)
+
+    if (NEKO_BCKND_DEVICE .eq. 1) then
+       call device_col2(s%x_d, coef%mult_d, s%dof%size())
+    else
+       call col2(s%x, coef%mult, s%dof%size())
```
Comment (timofeymukha) [inline src/scalar/scalar_ic.f90:109]: Just a question, what does this do? I looked in coef and mult seems to be set to 1.

File: src/scalar/scalar_ic.f90 [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+module scalar_ic
+  use gather_scatter
+  use neko_config
+  use num_types
+  use device_math
+  use device
+  use field
+  use utils
+  use coefs
+  use math
```
Comment (timofeymukha) [inline src/scalar/scalar_ic.f90:43]: Please use `only` and in these as well as much as possible.

File: src/scalar/scalar_ic.f90 [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    type(coef_t), intent(in) :: coef
+    type(gs_t), intent(inout) :: gs
+    character(len=*) :: type
+    type(json_file), intent(inout) :: params
+    ! Variables for retrieving json parameters
+    real(kind=rp) :: ic_value
+
+    if (trim(type) .eq. 'uniform') then
+       call json_get_or_default(params, 'case.scalar.initial_condition.value', &
+                                ic_value, 0.0_rp)
```
Comment (timofeymukha) [inline src/scalar/scalar_ic.f90:70]: I would say, no need for a default here. A value of zero is not really special in any way to default to it.

File: reframe/src/rayleigh.f90 (hunk 1 above)
Comment (njansson) [reply reframe/src/rayleigh.f90:72]: Ok great!

File: src/case.f90 [hunk 5]
```diff
@@ -276,6 +277,18 @@ subroutine case_init_common(C)
             C%fluid%c_Xh, C%fluid%gs_Xh, C%usr%fluid_user_ic, C%params)
     end if
 
+    if (scalar) then
+       call json_get_or_default(C%params, 'case.scalar.initial_condition.type',&
+                  string_val, 'uniform')
```
Comment (timofeymukha) [inline src/case.f90:282]: Whether to have a default here is also a question. We have no default for fluid, and I think it is best to be consistent. I think initial conditions is such an important thing that it is best to have it completely explicit.

File: src/scalar/scalar_ic.f90 [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+       call set_scalar_ic_uniform(s, ic_value)
+    else
+       call neko_error('Invalid initial condition')
+    end if
+
+    call set_scalar_ic_common(s, coef, gs)
+
+  end subroutine set_scalar_ic_int
+
+  !> Set scalar intial condition (user defined)
```
Comment (timofeymukha) [inline src/scalar/scalar_ic.f90:80]: Please add `@param` doctstrings for the dummy arguments here and in the other subroutines.

**Review by njansson — APPROVED** [review]

**Review by timofeymukha — CHANGES_REQUESTED** [review]

**Comment by timofeymukha** [comment]
Note: If we agree to not drop back to uniform 0 as a default, as I suggest, we need to update the case files of all the examples using `scalar`.

File: src/scalar/scalar_ic.f90 (hunk 2 above)
Comment (timfelle) [reply src/scalar/scalar_ic.f90:109]: > Just a question, what does this do? I looked in coef and mult seems to be set to 1.

I have no idea. 
It is copied directly from flow_ic.f90, so I figured it was important.

**Comment by timfelle** [comment]
> Note: If we agree to not drop back to uniform 0 as a default, as I suggest, we need to update the case files of all the examples using `scalar`.

Yes, which is exactly why i thought it might be useful to have the default. It is the current behavior and would avoid breaking examples other people might be working with. But yes, it is a disconnect from the fluid way of doing initial conditions.

File: src/scalar/scalar_ic.f90 (hunk 3 above)
Comment (timfelle) [reply src/scalar/scalar_ic.f90:43]: Done, will wait with a push till we have resolved the other comments as well.

File: src/scalar/scalar_ic.f90 (hunk 6 above)
Comment (timfelle) [reply src/scalar/scalar_ic.f90:80]: Done, will wait with a push till we have resolved the other comments as well.

File: src/case.f90 (hunk 5 above)
Comment (timfelle) [reply src/case.f90:282]: Lets keep it consistent. It makes the most sense, and I guess its not going to be that much of a bother for people.

Comment (timofeymukha) [reply src/case.f90:282]: Yes. @adperezm since you are the main scalar user, here is a heads up!

File: src/scalar/scalar_ic.f90 (hunk 4 above)
Comment (timfelle) [reply src/scalar/scalar_ic.f90:70]: with the default initial condition type removed, this is also removed.

**Pull request merged**